### DISCUSSION TITLE
Static interactive plot by running a bokeh server in webassembly using pyodide

### DISCUSSION THREAD

**Discussion by choucavalier** [bokeh/bokeh#11693] [General]
I see a very interesting use-case in serving an interactive bokeh plot within a static web page, with the bokeh server running in websassembly via pyodide ( https://github.com/pyodide/pyodide ).

One example of application would be to have interactive plots, with Python callbacks, within a statically-generated website (e.g. a scientific blog). Because everything would run in the web browser, even the bokeh server, this would scale much better than having to run a bokeh server answering requests for all visitors.

It seems that bokeh is supported by pyodide ( https://github.com/tytgatlieven/pyodide/commit/364581473e2c7375502743b94f3737cd49ec79a5 ).

If this is already possible, maybe it would be a good idea to provide an example in the bokeh documentation.

In particular, on this page of the documentation https://docs.bokeh.org/en/latest/docs/user_guide/embed.html it says that the standalone document does not require a bokeh server to run. If this webassembly-ran bokeh server idea is possible right now using pyodide, then we can **extend** this definition of standalone documents because we could potentially run standalone documents that do run a bokeh server.

I'm opening the discussion here about this subject, and I'm potentially missing something, because I'm no expert in this new tech. However, my internet search did not lead to me to a conclusive answer on this subject.

What do you think?

**Comment by bryevdv** [comment]
@tgy Using the Bokeh server requires Tornado, which is not included in the requirements in the recipe you link above, that means that Bokeh would be limited exclusively to standalone (non-Bokeh server) capabilities. AFAIK porting Tornado to pyodide would be huge undertaking, and I have not heard of anyone claiming to be working on it, so I would not expect Bokeh server in its current form to use usable on pyodide any time soon. 

Some other possibilities:

* I would like to make the Bokeh synchronization protocol usable without Tornado. This could potentially open up an avenue for running something very similar to the current Bokeh server on pyodide. But this is also a heavy lift, on the Bokeh side. Unless the work is  finds support in some way I would not expect it to happen any time soon.

* Another purely hypothtical possibility might be if Python callbacks could be compiled to something that could be used with `js_on_change` etc, bypassing the need for the Bokeh server altogether in the specific context of pyodide. We used to have something vaguely similar to the with `CustomJS.from_py_func` that did a Python side transpilation to JS. It was very limited (only a subset of python stdlib, no numpy etc) and not very robust so we eventually dropped it entirely. But If there was a way to call out to numpy and pandas etc, that would make this much more interesting and potentially worth trying. But experiments with that would be greatly accelerated if someone with pyodide/wasm experience decides to take lead.

**Reply by choucavalier** [reply]
Thank you for your detailed answer!

**Reply by MarcSkovMadsen** [reply]
What about using "Voila" server in jupyterlite to serve Bokeh/ Panel is that an "option".

You should be able to serve Bokeh/ Panel widgets and apps using Voila server normally.

**Reply by philippjfr** [reply]
Certainly Panel can't currently be imported without installing Tornado. But I think Bokeh+jupyter_bokeh.BokehModel might just work inside jupyterlite.
